google/gax/api_callable.py

```diff
@@ -306,8 +293,64 @@ def _upper_camel_to_lower_under(string):
     return out
 
 
+def _override_retry(config_base, overrides):
+    """Update config_base specifying retry params with overrides.
+
+    Args:
+      config_base: A dictionary of retry params.
+      overrides: A dictionary to update config_base.
+    """
+    for (name, params) in overrides.items():
+        if name in config_base:
+            config_base[name].update(params)
+        else:
+            config_base[name] = params
+
+
+def _override_methods(config_base, overrides):
+    """Update config_base specifying method condigs with overrides.
+
+    Args:
+      config_base: A dictionary of method config.
+      overrides: A dicitionary to update config_base.
+    """
+    for (name, params) in overrides.items():
+        if not params:
+            config_base[name] = None
+        else:
+            for (key, param) in params.items():
+                if key not in config_base[name]:
+                    continue
+                if key == 'bundling' and param:
+                    config_base[name][key].update(param)
+                else:
+                    config_base[name][key] = param
+
+
+def _override_config(config_base, overrides):
+    """Merge overriding configs into config_base and return the new config.
+
+    If overrides is empty, it returns the same config_base. Otherwise,
+    it creates a copy of config_base and some parts are updated by overrides.
+
+    Args:
+      config_base: A dictionary for base config.
+      overrides: A dictionary in the same structure of config_base.
+    """
+    if not overrides:
+        return config_base
+    new_config = copy.deepcopy(config_base)
+    if 'retry_codes' in overrides:
```

Comment:
Just a thought -- is this really the behavior we want if a user redefines a `retry_code` or `retry_params`? For example, in Pubsub, there is a [`messages`](https://github.com/googleapis/googleapis/blob/master/google/pubsub/v1/pubsub_gapic.yaml#L41) `retry_params`. Isn't it confusing if I define a new `messages` retry parameter set without realizing that there's already one with that name, and then accidentally override a bunch of method settings?

What if we raise an exception in this case and say that the name is reserved? Or even better, ensure that a user-provided parameter name is handled distinctly from a default parameter of the same name?